pkg/diagnose/connectivity/hooks.go

```diff
@@ -0,0 +1,118 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package connectivity contains logic for connectivity troubleshooting between the Agent
+// and Datadog endpoints. It uses HTTP request to contact different endpoints and displays
+// some results depending on endpoints responses, if any.
+package connectivity
+
+// This file contains all the functions used by httptrace.ClientTrace.
+// Each function is called at a specific moment during the communication
+// Their prototypes are defined by htpp.Client so variables might be unused
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http/httptrace"
+
+	"github.com/fatih/color"
+)
+
+// During a request, the http.Client will call the functions of the ClientTrace at specific moments
+// This is useful to get extra information about what is happening and if there are errors during
+// connection establishment, DNS resolution or TLS handshake for instance
+var DiagnoseTrace = &httptrace.ClientTrace{
+
+	// Hooks called before and after creating or retrieving a connection
+	GetConn: getConnHook,
+	GotConn: gotConnHook,
+
+	// Hooks for connection establishment
+	ConnectStart: connectStartHook,
+	ConnectDone:  connectDoneHook,
+
+	// Hooks for DNS resolution
+	DNSStart: dnsStartHook,
+	DNSDone:  dnsDoneHook,
+
+	// Hooks for TLS Handshake
+	TLSHandshakeStart: tlsHandshakeStartHook,
+	TLSHandshakeDone:  tlsHandshakeDoneHook,
+}
+
+var (
+	EmptyTrace   = &httptrace.ClientTrace{}
+	dnsColorFunc = color.MagentaString
+	tlsColorFunc = color.YellowString
+)
+
+// connectStartHook is called when the http.Client is establishing a new connection to 'addr'
+// However, it is not called when a connection is reused (see gotConnHook)
+func connectStartHook(network, addr string) {
+	fmt.Printf("~~~ Starting a new connection ~~~\n")
+}
+
+// connectDoneHook is called when the new connection to 'addr' completes
+// It displays the error message if there is one and indicates if this step was successful
+func connectDoneHook(network, addr string, err error) {
+	statusString := color.GreenString("OK")
+	if err != nil {
+		statusString = color.RedString("KO")
```

Comment:
I saw "KO" for the first time in the DR exercise last week, and figured out from context that it was used to mean "not OK".  I haven't seen it before -- maybe it means "knock-out"?  Or maybe it's a reference to OK K.O.?  Anyway, I suspect in this context it will not be clear what it means -- maybe "OK" and "FAIL" would be better?

Reply:
I'd +1 on this as well. Either `PASS`/`FAIL` or `OK`/`FAIL` are probably the most widely expected combinations here.

Reply:
Done. I used `ERROR` instead of `FAIL` to emphasize the difference between small steps and full pipeline

- For small steps (ssl, tls, ...) : `OK / ERROR` 
- For the full pipeline : `PASS / FAIL`

Let me know if it is legit for you, otherwise I can use `FAIL` instead of `ERROR`